From 94de79787f1247174918182d1f493d4af0ed74d7 Mon Sep 17 00:00:00 2001
From: Tim Hartley <timothy.d.hartley@intel.com>
Date: Fri, 19 Jul 2013 09:24:55 +0000
Subject: Dalvik: Build PCG-based JIT compiler backend by default

BZ: 124256

This patch enables building libpcgdvmjit.so, the PCG-based Dalvik JIT
compiler backend, when building the default build target. Thus, when
executing, for instance, 'make redhookbay', libpcgdvmjit.so will
be built for the target and installed into the appropriate
/system/lib/ dir.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG; AOSP-Dalvik-Runtime
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: Iee7291bf59d40f1fe427fa09b230407ce2e5e845
Orig-MCG-Change-Id: I2f937508c0c3dccf03f5ed48e031757edcdd3c76
Signed-off-by: Tim Hartley <timothy.d.hartley@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Reviewed-on: http://android.intel.com:8080/124378
Reviewed-by: Lupusoru, Razvan A <razvan.a.lupusoru@intel.com>
Reviewed-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Tested-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 Android.mk                             |   31 +++++++++++++++++++++----------
 vm/Android.mk                          |    7 +++++++
 vm/compiler/codegen/x86/pcg/Android.mk |    2 +-
 3 files changed, 29 insertions(+), 11 deletions(-)

diff --git a/Android.mk b/Android.mk
index 12543ff..0fa8745 100644
--- a/Android.mk
+++ b/Android.mk
@@ -19,20 +19,31 @@ ifeq ($(WITH_JIT),)
   ifeq ($(TARGET_ARCH),x86)
     WITH_JIT := true
 
-  # For all builds, except for the -user build we will enable support for
-  # VTune Amplifier by default. To collect any information about JIT look
-  # at -Xjitvtuneinfo:<info> option. By default, -Xjitvtuneinfo:none is
-  # used and it doesn't affect JIT compiler (one extra 'if' per compiled
-  # trace).
-  ifneq ($(TARGET_BUILD_VARIANT),user)
-    VTUNE_DALVIK := $(strip $(VTUNE_DALVIK))
-    ifeq ($(VTUNE_DALVIK),)
-      VTUNE_DALVIK := true
+    # For all builds, except for the -user build we will enable support for
+    # VTune Amplifier by default. To collect any information about JIT look
+    # at -Xjitvtuneinfo:<info> option. By default, -Xjitvtuneinfo:none is
+    # used and it doesn't affect JIT compiler (one extra 'if' per compiled
+    # trace).
+    ifneq ($(TARGET_BUILD_VARIANT),user)
+      VTUNE_DALVIK := $(strip $(VTUNE_DALVIK))
+      ifeq ($(VTUNE_DALVIK),)
+        VTUNE_DALVIK := true
+      endif
     endif
-  endif
 
   else
     WITH_JIT := false
+    WITH_PCG := false
+  endif
+endif
+
+# Turning on WITH_PCG when WITH_JIT=true (either through command-line
+# or setting it to true above), and when WITH_PCG isn't already set
+ifeq ($(WITH_JIT),true)
+  ifeq ($(TARGET_ARCH),x86)
+    ifeq ($(WITH_PCG),)
+        WITH_PCG := true
+    endif
   endif
 endif
 
diff --git a/vm/Android.mk b/vm/Android.mk
index 61fea0f..bd5aa38 100644
--- a/vm/Android.mk
+++ b/vm/Android.mk
@@ -53,6 +53,13 @@ endif
 
 # Overwrite default settings
 LOCAL_MODULE := libdvm
+ifeq ($(WITH_JIT),true)
+  ifeq ($(TARGET_ARCH),x86)
+    ifeq ($(WITH_PCG),true)
+      LOCAL_REQUIRED_MODULES += libpcgdvmjit
+    endif
+  endif
+endif
 LOCAL_CFLAGS += $(target_smp_flag)
 
 # Define WITH_ADDRESS_SANITIZER to build an ASan-instrumented version of the
diff --git a/vm/compiler/codegen/x86/pcg/Android.mk b/vm/compiler/codegen/x86/pcg/Android.mk
index f5dc00f..ec38892 100644
--- a/vm/compiler/codegen/x86/pcg/Android.mk
+++ b/vm/compiler/codegen/x86/pcg/Android.mk
@@ -42,7 +42,7 @@ LOCAL_CFLAGS += $(target_smp_flag)
 LOCAL_CFLAGS += -DARCH_IA32 -DWITH_JIT -DWITH_HOUDINI -DWITH_REGION_GC -DVTUNE_DALVIK -DEXTRA_SCRATCH_VR -DMTERP_STUB -DMTERP_NO_UNALIGN_64 -Wall -Wextra -O3
 LOCAL_ALLOW_UNDEFINED_SYMBOLS := true
 TARGET_LIBGCC = $(shell $(TARGET_CC) -m32 -print-libgcc-file-name)
-LOCAL_LDFLAGS += -Wl,--whole-archive $(TARGET_LIBGCC) -Wl,--no-whole-archive
+#LOCAL_LDFLAGS += -Wl,--whole-archive $(TARGET_LIBGCC) -Wl,--no-whole-archive
 LOCAL_STATIC_LIBRARIES += libpcg libirc libsvml libgnustl_static
 LOCAL_SRC_FILES := Analysis.cpp \
     Client.cpp \
-- 
1.7.4.1

